### DISCUSSION TITLE
Imports are not being transformed after hitting "return [something]" in parser.hooks.importSpecifier.tap

### DISCUSSION THREAD

**Discussion by TillThe** [webpack/webpack#13929] [Q&A]
So the issue was already actually mentioned there https://github.com/webpack/webpack/issues/12409 and there https://github.com/webpack/webpack/issues/12410. I suppose it's a bug, but I want to confirm if there's no solution. Problem is that imports are not being transformed after hitting `return [something]` in `parser.hooks.importSpecifier.tap()`

```
const importUseTag = Symbol('my tag');
parser.hooks.importSpecifier
      .tap(this.name, (expr, pathName, exportName, name) => {
        const request = parser.state.module.request;
        const ids = exportName === null ? [] : [exportName];

        if (pathName.endsWith([pathToL10n])) {
          l10nMap.set(request, {
            name,
            pathName
          });

          parser.tagVariable(name, importUseTag, {
            name,
            source: pathName,
            ids,
            sourceOrder: parser.state.lastHarmonyImportOrder,
            await: expr.await
          });
          return true; // parser.hooks.call is being called only with this
        }
      });

parser.hooks.call
      .for(importUseTag)
      .tap(this.name, (expr) => {
         // Exploring resources and args where l10n import is called
      });
```

Source code:

```
import { l10n } from "right_path_to_l10n";

const pkg = l10n('some_package');

```

As the result in bundle we have this:

```
/* harmony import */ var core_l10n_l10n__WEBPACK_IMPORTED_MODULE_10__ = __webpack_require__(/*! right_path_to_l10n */ 68874); 

var pkg = l10n('some_package'); // using undefined l10n instead of core_l10n_l10n__WEBPACK_IMPORTED_MODULE_10__

```

**Comment by matzeeable** [comment]
Hey @TillThe !

> As the result in bundle we have this:

Yeah, as soon as you `return true` in the `parser.hooks.importSpecifier.tap` callback, the imported function will be transformed to a "global" function call. I do not know if this is the intended behavior. After the transformation, it is totally correct that `parser.hooks.call.for(importUseTag)` will be triggered as your imported function is now globally available. Refer to this: https://webpack.js.org/api/parser/#call

In most cases, this will lead to `Uncaught ReferenceError: l10n is not defined`

@alexander-akait is this a bug? Or is there a way to tap into imported function calls **without transforming**?

> Called when parsing a function call.

@TillThe Did you already found a solution/workaround for this?

**Reply by alexander-akait** [reply]
I am afraid it was not design for this...Can you create minimum reproducible example, it will be more easy to say how achieve it

**Reply by matzeeable** [reply]
@alexander-akait Please fork https://codesandbox.io/s/webpack-webpack-discussions-13929-8x6nf?file=/src/scripts/index.js and there you will see that I have executed two function calls: 1x `__` and 1x `_n`. The file `webpack/webpack.config.dev.js` has a custom plugin `MyCustomWebpackPlugin`, and unfortunately, the `parser.hooks.call.for` does only work for `_n` as it is used as a global (**not imported!**) variable. How can we catch expressions for the imported function `__` ?

**Reply by alexander-akait** [reply]
I have solution, but I think you should not use it:
```
const HarmonyImportDependencyParserPlugin = require("webpack/lib/dependencies/HarmonyImportDependencyParserPlugin.js");
parser.hooks.callMemberChain.for(HarmonyImportDependencyParserPlugin.harmonySpecifierTag).tap(plugin, (expr) => {
  const { callee } = expr;
          
  if (callee.name === "__") {
    return false;
  }
});
```

webpack implements a lot of things for `import`/`export`, overriding logic and taking into account all possible nuances will not be so easy. Also you can see we import non exported file from webpack, because as I describe about it is dirty solution and webpack was not design for this. Why do not use babel/swc? Anyway you can implement own dependencies on `__`/`_n` and do not use `import`. And why do you need import?

**Reply by matzeeable** [reply]
> overriding logic and taking into account all possible nuances will not be so easy.

@alexander-akait in my case, I do not want to modify the AST in any way, I just want to extract the localization keys from the imported function. Thanks for the suggestion, I will give it a try. Perhaps it would be a good idea, to implement something like `parser.hooks.callImportedFunction.for`. What do you think?

> Why do not use babel/swc?

In my opinion, webpack is my single point of processing my files. In my case, I am currently using `babel`, but I have already `swc` on my todo list. Instead of rewriting plugins for JavaScript transformers, I want to benefit from the already created AST / ESTree of webpack and write the plugin once for all the same targets (`javascript/auto`, ...).

> And why do you need import?

I am using a factory function creating all my localization functions for **the current entry point.** They read the translated text on runtime instead of compile time. I want to avoid global modules and have all coding scoped to the module.

**Reply by alexander-akait** [reply]
> Perhaps it would be a good idea, to implement something like parser.hooks.callImportedFunction.for. What do you think?

hm, so you want to collect all possible variants from arguments?

**Reply by matzeeable** [reply]
Yes, I want to collect all translation arguments passed to the imported function `__` and write it to a POT file.

Another question regardless of the current issue... is it somehow possible to get the chunk filename where the parsed `CallExpression` gets put? From your previous solution it works fine to collect all function calls for `__`, and now I want to create a map link this:

```
[
  { "key": "Translate me 2", "file": "481.js" },
  { "key": "Translate me 3", "file": "923.js" }
  { "key": "Translate me 4", "file": "12.js" }
]
```

The `file` is the chunk name where the expression got found and for which webpack emits a file. Is this somehow possible?

**Reply by alexander-akait** [reply]
I think you mean original file path to module, because code generation is the latest step and we can generate chunk name only after we generated code (bundled), but unfortunately it is impossible/very hard, webpack parser get source code without filename https://github.com/webpack/webpack/blob/main/lib/javascript/JavascriptParser.js#L3290, in theory, of course, we can add additional properties to modules and use them, but it will be hacky, I am afraid we really was not design for visitors abilities

**Reply by matzeeable** [reply]
@alexander-akait Thanks again for your helpful answer. Ok, I think I could work around this as follow:

```js
const HarmonyImportDependencyParserPlugin = require("webpack/lib/dependencies/HarmonyImportDependencyParserPlugin.js");
parser.hooks.callMemberChain.for(HarmonyImportDependencyParserPlugin.harmonySpecifierTag).tap(plugin, (expr) => {
  const { callee } = expr;
          
  if (callee.name === "__") {
    return asSource(`/*THIS_GOT_FOUND:${moduleName}>*/${originalExpressionAsString}/*<THIS_GOT_FOUND*/`);
  }
});
```

In this case, the following things do not exist, or I do not exactly know, how I can get it within the function:

- `asSource()`: This function takes a string and returns the AST as replacement back, so the `THIS_GOT_FOUND` comment gets output to the bundle. I tried it with `acorn.Parser.parse` but this does not work with comments? How can I add a leading and trailing comment to the expression?
- `moduleName`: The original module name file (e.g. `my-file.tsx`)
- `originalExpressionAsString`: The original function call as string (e.g. `__("My test")`)

I hope I could clearly explain this. Do you have an idea how I could archive this?

**Reply by alexander-akait** [reply]
I am afraid there is no easy solution, we are really was not design for this, even we found hacky way, it may be broken for webpack v6
